rfcs/0004-persistent-ledger-builder-controller.md

```diff
@@ -0,0 +1,67 @@
+# Summary
+[summary]: #summary
+
+This adds a persistent backend for the ledger builder controller.
+
+# Motivation
+[motivation]: #motivation
+
+Having a persistent backend will allow nodes that go offline, whether for a short period or an extended period of time, to catch back up to the network with as little work as reasonably possible.
+
+# Detailed design
+[detailed-design]: #detailed-design
+
+In order to achieve this, we will persist the locked tip using the existing Merkle database. The backend Merkle database is designed in a way such that queries for the most commonly performed operations can be batched together as one request to the database.
+
+The best tip, materialized by the ledger builder controller, will be represented as a sparse Merkle ledger mask on top of the persistent Merkle ledger database. This data structure will only store the modified nodes of the Merkle ledger in memory, and defer to the contents of the persistent Merkle ledger database for any information outside of that. The persistent Merkle ledger database will maintain a reference to all of these data structures derived for it. These references will be used to garbage collect outdated information in the sparse diff representations whenever changes are written to the persistent copy. In other words, information will be removed from the sparse diff whenever it is no longer different from the information in the underlying database, ensuring that the data structure will not leak memory. If a sparse diff is empty, it may need to be retained, in case there are existing references to it.
+
+There may be several points between the locked tip and the best tip, and we may need to examine any such point. Most often, though, we'll want the diff between those extrema, and not for intermediate points. To reach intermediate points, we can replay individual operations from the locked tip (perhaps caching that diff). It's possible to create a series of diffs, or their composition, to reach intermediate points, but diffs will consume considerable memory. Since accessing intermediate points should be relatively rare, we'll maintain just a single diff.
+
+It's useful if there are operations on diffs with nice mathematical properties. For example, there could be an addition operation on diffs that results in the composition of those diffs. To "undo" a diff, one could produce the negation of a diff. These ideas suggests an algebraic group, where the carrier set is the set of diffs. The OCaml representation might be a Map from account keys or public keys to Amount.t could be sufficient to implement such a group discipline. An alternative would be a prefix tree with hashes summarizing accounts that have not changed, so that the root is a valid Merkle hash.
```

Comment:
We don't have those algebraic operations in the implementation. Do we need them?

Reply:
we could move this part to a different section of the RFC just in case we end up needing this behavior in the future